src/util/MathUtil.h

```diff
@@ -6,10 +6,30 @@
 #include <sstream>
 #include <vector>
 
+#include "boost/tuple/tuple.hpp"
+
 class MathUtil {
  public:
   static double Round(double x, int decimals = 0);
-  static bool AlmostEqual(double x, double y, int precision);
+  static bool AlmostEqual(double x, double y, int precision = 6);
+
+  // added by Yang
+  static int factorial(int n);
+  static double LogNChooseK(int n, int k);
+  static double LogSumExp(std::vector<double>* log_values); // en.wikipedia.org/wiki/LogSumExp
+
+  static double MaxInArr(double* arr_values, int size);
+  static double NormalizedDotProduct(double* src_values, double* tgt_values, int size, bool take_sqrt=true);
+  static std::vector<double> linspace(double start, double end, int num);
+
+  // find the closest match index from the data array to the query in O(log(n))
+  static int binarySearch(const double* data_arr, int data_size, double query);
```

Comment:
I'm surprised this isn;t implmeneted somewhere else already and you had to do it yourself.